pennylane/transforms/condition.py

```diff
@@ -93,8 +102,72 @@ def qnode():
             return qml.expval(qml.PauliZ(1))
     """
 
-    @wraps(then_op)
-    def wrapper(*args, **kwargs):
-        return Conditional(measurement, then_op(*args, do_queue=False, **kwargs))
+    if inspect.isclass(then_func) and issubclass(then_func, Operator):
+
+        @wraps(then_func)
+        def wrapper(*args, **kwargs):
+            ops = []
+            then_cond_op = Conditional(measurement, then_func(*args, do_queue=False, **kwargs))
+            ops.append(then_cond_op)
+
+            if else_func:
+
+                # Copy the MV such that the external state is not changed
+                inverted_m = copy(measurement)
+                else_cond_op = Conditional(~inverted_m, else_func(*args, do_queue=False, **kwargs))
+                ops.append(else_cond_op)
+
+            return ops
+
+    elif callable(then_func):
+
+        with_meas_err = (
+            "Only quantum functions that contain no measurements can be applied conditionally."
+        )
+
+        if else_func is None:
+
+            @wraps(then_func)
+            def wrapper(*args, **kwargs):
+                # We assume that the callable is a quantum function
+                tape = make_tape(then_func)(*args, **kwargs)
+
+                if tape.measurements:
+                    raise ConditionalTransformError(with_meas_err)
+
+                for op in tape.operations:
+                    Conditional(measurement, op)
+
+        else:
+
+            @wraps(then_func)
+            def wrapper(*args, **kwargs):
+                # We assume that the callable is a quantum function
+
+                # 1. Apply then_func conditionally
+                tape = make_tape(then_func)(*args, **kwargs)
+
+                if tape.measurements:
+                    raise ConditionalTransformError(with_meas_err)
+
+                for op in tape.operations:
+                    Conditional(measurement, op)
+
+                # 2. Apply else_func conditionally
+                else_tape = make_tape(else_func)(*args, **kwargs)
+
+                if else_tape.measurements:
+                    raise ConditionalTransformError(with_meas_err)
+
+                inverted_m = copy(measurement)
+                inverted_m = ~inverted_m
+
+                for op in else_tape.operations:
+                    Conditional(inverted_m, op)

```

Comment:
It looks like this could also be simplified to just
```python
@wraps(then_func)
def wrapper(*args, **kwargs):
    # We assume that the callable is a quantum function

    # 1. Apply then_func conditionally
    tape = make_tape(then_func)(*args, **kwargs)

    if tape.measurements:
        raise ConditionalTransformError(with_meas_err)

    for op in tape.operations:
        Conditional(measurement, op)

    if else_func is not None:
        # 2. Apply else_func conditionally
        else_tape = make_tape(else_func)(*args, **kwargs)
    
        if else_tape.measurements:
            raise ConditionalTransformError(with_meas_err)
    
        inverted_m = copy(measurement)
        inverted_m = ~inverted_m
    
        for op in else_tape.operations:
            Conditional(inverted_m, op)
```

Reply:
That is, moving the `if else_func is not None:` _inside the wrapper_.

Reply:
This single code block should cover all edge cases:

- qfunc, no else: `qml.cond(m == 0, qfunc)`
- qfunc, with else: `qml.cond(m == 0, qfunc1, qfunc2)`
- operator, no else: `qml.cond(m == 0, qml.RX)`
- operator, with else: `qml.cond(m == 0, qml.RX, qml.RY)`

Reply:
Oh, that's right! This works like a charm (changed in https://github.com/PennyLaneAI/pennylane/pull/2275/commits/58f531acfbc91e3ed50d84843f80ecec9244675d), so nice to have this simplified. Thanks for the suggestion!